File: DP/counting-tiles.cpp
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

const int mod = 1000000007;

void generate_next_masks(int current_mask, int i, int next_mask, int n, 
                            vector<int>& next_masks){
      if(i == n + 1){
        next_masks.push_back(next_mask);
        return;
      }
      
      if((current_mask & (1 << i)) != 0)
          generate_next_masks(current_mask, i + 1, next_mask, n, next_masks);
      
      if(i != n)
        if((current_mask & (1 << i)) == 0 && (current_mask & (1 << (i+1))) == 0)
          generate_next_masks(current_mask, i + 2, next_mask, n, next_masks);      
      
      if((current_mask & (1 << i)) == 0)
          generate_next_masks
                (current_mask, i + 1, next_mask + (1 << i), n, next_masks);    
}
 
int dp[1001][1<<11];
int solve(int col, int mask, const int m, const int n){
    // BASE CASE
    if(col == m + 1){
        if(mask == 0)
          return 1;
        return 0;
    }
 
    if(dp[col][mask] != -1)
      return dp[col][mask];
 
    int answer = 0;
    vector<int> next_masks;
    generate_next_masks(mask, 1, 0, n, next_masks);
 
    for(int next_mask: next_masks){
        answer = (answer + solve(col + 1, next_mask, m, n)) % mod;
    }
 
    return dp[col][mask] = answer;
}
 
int main() {
   int t = 1; //cin >> t;
   while(t--){
        int n, m;
        cin >> n >> m;
        memset(dp, -1, sizeof dp);
        cout << solve(1, 0, m, n); 
   }
   return 0;
}

File: DP/counting-tiles.py
# NOT WORKING

n, m = map(int, input().split())

dp = [[-1 for _ in range(1<<n)] for _ in range(m)]
MOD = 1000000007

def generate_next_masks(row, mask, next_mask, next_masks):
    if row == n:
        next_masks.append(next_mask)
        return

    if mask & 1<<row != 0:
        generate_next_masks(row+1, mask, next_mask, next_masks)
    
    if row != n-1 and (mask & 1<<row == 0) and (mask & 1<<row+1 == 0):
        generate_next_masks(row+2, mask, next_mask, next_masks)

    if (mask & 1<<row == 0):
        generate_next_masks(row+1, mask, next_mask | 1<<row, next_masks)  # Make a copy here

    return

def solve(col, mask):
    if col == m:
        if mask == 0:
            return 1
        return 0

    if dp[col][mask] != -1:
        return dp[col][mask]

    ans = 0

    next_masks = []

    generate_next_masks(0, mask, 0, next_masks)

    for next_mask in next_masks:
        ans = (ans + solve(col+1, next_mask)) % MOD

    dp[col][mask] = ans

    return dp[col][mask]

# current_col, mask
print(solve(0, 0))
